ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs

```diff
@@ -304,46 +365,57 @@ toRealPoint (Weaken ap)      = toRealPoint ap
 --
 -- We take in the entire 'LedgerDB' because we record that as part of errors.
 applyBlock :: forall m c l blk
-            . (ApplyBlock l blk, TickedTableStuff l, Monad m, c)
+            . (ApplyBlock l blk, TickedTableStuff l, Monad m, c, HeaderHash blk ~ HeaderHash (l ValuesMK))
            => LedgerCfg l
            -> Ap m l blk c
            -> LedgerDB l
-           -> m (l ValuesMK, l TrackingMK)
-applyBlock cfg ap db = case ap of
+           -> (l ValuesMK -> LedgerDB l -> LedgerDB l)
+           -> (l TrackingMK -> LedgerDB l -> LedgerDB l)
+           -> m (LedgerDB l) --m (l ValuesMK, l TrackingMK)
+applyBlock cfg ap db pushOld pushNew = case ap of
     Weaken ap' ->
-        applyBlock cfg ap' db
-    _ -> do
-      old <- oldApplyBlock ap
-      new <- newApplyBlock ap
-
-      assert (old == applyTracking oldLedgerDbCurrent new) $ return (old, new)
+        applyBlock cfg ap' db pushOld pushNew
+    _ -> case runDual db of
+      RunOld _ -> do
+        old <- oldApplyBlock ap
+        case ledgerDbPastDiffs (castPoint @(l ValuesMK) @blk $ getTip old) db of
+          Nothing   -> error "New database doesn't have point resulting frome executing old database"
+          Just past -> assert (old == applyDiffs (ledgerDbCurrentOld db) past) $ return $ pushOld old db
```

Comment:
This is wrong, there is no information before the UTxO disk backend and therefore there might be cases in which `Nothing` is returned